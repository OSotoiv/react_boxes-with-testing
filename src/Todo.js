import React, { useState } from "react";


const Todo = ({ task, removeTodo, state, id }) => {
    const [isHovered, setIsHovered] = useState(false);
    const handleMouseEnter = () => {
        setIsHovered(true);
    };

    const handleMouseLeave = () => {
        setIsHovered(false);
    };
    return (
        <li data-testid="todo" className="todo"
            onMouseEnter={handleMouseEnter}
            onMouseLeave={handleMouseLeave}
            style={{
                textDecoration: `${task.state === 'done' ? 'line-through' : ''}`
            }}
        >
            {task.todo.toUpperCase()}
            {isHovered && task.state === 'incomplete' ?
                <button data-testid="todo-remove" className="todo-remove" type='button' onClick={() => state(id)}>done</button> : null}
            {isHovered && <button data-testid="todo-remove" className="todo-remove" type='button' onClick={() => removeTodo(id)}>X</button>}
        </li >)
}
export default Todo;