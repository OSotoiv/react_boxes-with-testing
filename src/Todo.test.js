import React from "react";
import Todo from './Todo'
import { render, fireEvent } from "@testing-library/react";

const todo = {
    id: 'todo-123',
    todo: 'Write Testing',
    state: 'incomplete',
    removeTodo: function (e) { e.preventDefault() },
    isDone: function (e) { e.preventDefault() }
}

it('render SMOKE TEST', () => {
    render(<Todo task={todo} removeTodo={todo.removeTodo} isDone={todo.isDone} id={todo.id} key={todo.id} />)
});

it('matches snapshot', () => {
    const { asFragment } = render(<Todo task={todo} removeTodo={todo.removeTodo} isDone={todo.isDone} id={todo.id} key={todo.id} />)
    expect(asFragment()).toMatchSnapshot();
});

it('renders li for a todo', () => {
    const { queryByTestId } = render(<Todo task={todo} removeTodo={todo.removeTodo} isDone={todo.isDone} id={todo.id} key={todo.id} />)
    const li = queryByTestId('todo');
    expect(li).toBeInTheDocument();
    expect(li.innerHTML).toBe(todo.todo.toUpperCase())
});
it('renders a done button and a delete button when mouseover the todo', () => {
    const { queryByTestId } = render(<Todo task={todo} removeTodo={todo.removeTodo} isDone={todo.isDone} id={todo.id} key={todo.id} />)
    const li = queryByTestId('todo');
    //done button and remove button are not visible
    expect(queryByTestId('is-done')).toBeNull();
    expect(queryByTestId('todo-remove')).toBeNull();

    //buttons are now visible during mouseover
    fireEvent.mouseEnter(li);
    expect(queryByTestId('is-done')).toBeInTheDocument();
    expect(queryByTestId('todo-remove')).toBeInTheDocument();

    //buttons are not visible when mouse leaves
    fireEvent.mouseLeave(li);
    expect(queryByTestId('is-done')).toBeNull();
    expect(queryByTestId('todo-remove')).toBeNull();

    //clicking a button should update the state of the list of todos
    //but we dont have the function on the todo itself unless it is passed down
    //so we can not test clicking the button in here.
});