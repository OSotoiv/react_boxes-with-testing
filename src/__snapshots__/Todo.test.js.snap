// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <li
    class="todo"
    data-testid="todo"
  >
    WRITE TESTING
  </li>
</DocumentFragment>
`;
